content/client-lib-development-guide/features.textile

```diff
@@ -129,6 +142,16 @@ h3(#rest-channel). Channel
 ** @(RSL1c)@ When an array of @Message@ objects is provided, a single request is made to Ably. When publishing multiple messages, this approach is more efficient. However, a yet to be implemented feature should limit the total number of messages bundled in a single POST based on the default max request size, and would raise an exception if any single message exceeds that limit.
 ** @(RSL1d)@ Indicates an error if the message was not successfully published to Ably
 ** @(RSL1e)@ Allows @name@ and or @data@ to be @null@.  If any of the values are null, then key is not sent to Ably i.e. a payload with a @null@ value for @data@ would be sent as follows @{ "name": "click" }@
+** @(RSL1f)@ Unauthenticated clients without a @clientId@ (any @clientId@ is permitted):
+*** @(RSL1f1)@ When a @Message@ with a @clientId@ value is published, Ably will accept and publish that message with the provided @clientId@. A test should assert via the history API that the @clientId@ of the published @Message@ is correct for both authentication schemes (token and basic)
+** @(RSL1g)@ Authenticated clients with a @clientId@ (as a result of either an explicitly configured @clientId@ in @ClientOptions@, or implicitly through token auth):
+*** @(RSL1g1)@ When publishing a @Message@ with the @clientId@ attribute set to @null@:
+**** @(RSL1g1a)@ It is unnecessary for the client to set the @clientId@ of the @Message@ before publishing
+**** @(RSL1g1b)@ Ably is expected to automatically assign a @clientId@ upon receiving the @Message@. A test should assert via the history API that the @clientId@ value is populated for the @Message@ when received
+*** @(RSL1g2)@ When publishing a @Message@ with the @clientId@ attribute value set to the authenticated client's @clientId@, Ably will accept the message and publish it. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** @(RSL1g3)@ When publishing a @Message@ with a different @clientId@ attribute value to the authenticated @clientId@, the client library should reject publishing immediately. The message should not be sent to Ably and an exception should be raised. The connection and channel remains available for further operations
+*** @(RSL1g4)@ When using token auth with a literal token, an auth URL or an auth callback, until authentication has completed, the client library will not be constrained to a @clientId@. However, as authentication always occurs before the message is published in a synchronous REST library, an exception should be raised if publishing a message with an incompatible @clientId@ to the @clientId@ that is set following authentication. The connection and channel remains available for further operations
```

Comment:
No, I meant the library. If you are using Token Auth because and have provided a `authCallback`, and are using the Realtime library, the Ably client library will let you publish a message with an incompatible `clientId` because at the time of publish there is no way for the client library to know what the clientId should be.  However, in the case of the REST client library, when publishing a message, it needs to call the `authCallback` and then execute the POST HTTP request to Ably.  However, before it even issues the POST request, it will know from the `TokenDetails` object that was returned from the `authCallback` that the `clientId` is incompatible, and can thus raise an exception without having tried to send it.